aiohttp_s3_client/credentials.py

```diff
@@ -0,0 +1,158 @@
+
+    def as_dict(self) -> dict:
+        return {
+            "region": self.region,
+            "access_key_id": self.access_key_id,
+            "secret_access_key": self.secret_access_key,
+            "service": self.service,
+        }
+
+
+class URLCredentials(Credentials):
+    def __init__(
+        self, url: Union[str, URL], *, region: str = "", service: str = "s3"
+    ):
+        url = URL(url)
+        super().__init__(
+            access_key_id=url.user or '',
+            secret_access_key=url.password or '',
+            region=region, service=service,
+        )
+
+
+class EnvironmentCredentials(Credentials):
+    def __init__(self, region: str = '', service: str = "s3"):
+        super().__init__(
+            access_key_id=os.getenv("AWS_ACCESS_KEY_ID", ''),
+            secret_access_key=(
+                os.getenv("AWS_SECRET_ACCESS_KEY", '') or
+                os.getenv("AWS_SESSION_TOKEN", '')
+            ),
+            region=os.getenv('AWS_DEFAULT_REGION', region),
+            service=service,
+        )
+
+
+class ConfigCredentials(Credentials):
+    DEFAULT_CONFIG_PATH = Path.home() / '.aws' / 'credentials'
+
+    def __init__(
+        self, config_path: Union[str, Path] = DEFAULT_CONFIG_PATH, *,
+        region: str = '', service: str = "s3", profile: str = 'auto'
+    ):
+        credentials_file = os.getenv('AWS_SHARED_CREDENTIALS_FILE')
+        if credentials_file is not None:
+            config_path = Path(credentials_file)
+
+        if not isinstance(config_path, Path):
+            config_path = Path(config_path)
+
+        if not config_path.exists():
+            super().__init__(
+                access_key_id='', secret_access_key='',
+                region=region, service=service,
+            )
+            return
+
+        if profile == 'auto':
+            profile = os.getenv('AWS_PROFILE', 'default')
+
+        creds = configparser.ConfigParser()
+        creds.read(config_path)
+
+        if profile not in creds:
+            super().__init__(
+                access_key_id='', secret_access_key='',
+                region=region, service=service,
+            )
+            return
+
+        section = creds[profile]
+
+        super().__init__(
+            access_key_id=section.get('aws_access_key_id'),
+            secret_access_key=section.get('aws_secret_access_key'),
+            region=section.get('region'), service=service,
+        )
+
+
+class MetadataCredentials(Credentials):
+    METADATA_ADDRESS = '169.254.169.254'
+    METADATA_PORT = 80
+
+    def _fast_probe(self):
+        with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as sock:
+            return sock.connect_ex(
+                (self.METADATA_ADDRESS, self.METADATA_PORT)
+            ) == 0
+
+    def _fetch(self, suffix: str) -> bytes:
+        metadata_url = (
+            f'http://{self.METADATA_ADDRESS}:{self.METADATA_PORT}/latest'
+        )
+        with urlopen(f'{metadata_url}{suffix}') as f:
```

Comment:
need a timeout=1 here with appropiate try except for urllib.error.URLError: <urlopen error timed out>

Reply:
resolved no timeout needed anymore )))